### PR TITLE
feat: manually implement write_all_vectored

### PR DESCRIPTION
**Pull request by PureWhiteWu**

### CODE REVIEW

**Review by ihciah — APPROVED** [review]

**Review by ihciah — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -77,40 +78,76 @@ impl LinkedBytes {
...
-    //     unsafe { &*(self.ioslice.as_mut_slice() as *mut _) }
-    // }
+        self.ioslice
+            .push(IoSlice::new(unsafe { &*(self.bytes.as_ref() as *const _) }));
+
+        // do write_all_vectored
+        let (mut base_ptr, mut len) = (self.ioslice.as_mut_ptr(), self.ioslice.len());
+        while len != 0 {
+            let ioslice = unsafe { std::slice::from_raw_parts_mut(base_ptr, len) };
+            let n = writer.write_vectored(ioslice).await?;
```
Comment (ethe) [inline src/lib.rs:108]: is it better to use `&self.ioslice[..]` instead?

File: src/lib.rs [hunk 2]
```diff
@@ -77,40 +78,76 @@ impl LinkedBytes {
...
+                *inner_slice =
+                    IoSlice::new(unsafe { std::slice::from_raw_parts(new_ptr, new_len) });
+            }
+        }
+        self.ioslice.clear();
+        Ok(())
+    }
     pub fn reset(&mut self) {
         // ioslice must be cleared before list
         self.ioslice.clear();
```
Comment (ethe) [inline src/lib.rs:153]: if the capacity of `self.ioslice` is too big, resize it to a default length should be better.

Comment (PureWhiteWu) [reply src/lib.rs:153]: I think the capacity can't be too big, since this is only used to contain IoSlice, not the inner buffer.

Comment (ethe) [reply src/lib.rs:153]: if we assume the capacity does not increase too much, I think `[IoSlice; FIXED_SIZE]` would be better than `Vec<IoSlice>`, it avoid an allocation in `LinkedBytes` initialization, and we are not able to assume how often `LinkedBytes` would be created.

File: src/lib.rs (hunk 1 above)
Comment (PureWhiteWu) [reply src/lib.rs:108]: I have changed the above line to from_raw_parts instead of from_raw_parts_mut

File: src/lib.rs (hunk 2 above)
Comment (ethe) [reply src/lib.rs:153]: such as https://github.com/hyperium/hyper/blob/89048f4054434435aa6eb4729d46787a89ad5857/src/proto/h1/io.rs#L300

Comment (PureWhiteWu) [reply src/lib.rs:153]: If we use fixed size, it will bring extra complexity to the implementation of write_all_vectored.

**Pull request merged**